src/utilities/pythonVariableLogger.py

```diff
@@ -0,0 +1,88 @@
+from typing import Callable, Any, Sequence, Union, Dict
+
+import numpy as np
+
+from Basilisk.architecture import sysModel
+
+LoggingFunction = Callable[[int], Any]
+
+class PythonVariableLogger(sysModel.SysModel):
+    """This a Python Module that will call one or multiple functions
+    and store their results. After the simulation is over, these
+    values can be retrieved.
+
+    This object is built with a dictionary of functions, and the
+    results can be retrieved from the object using the keys of
+    this dictionary::
+
+        log = VariableLogger({
+            "a": lambda CurrentSimNanos: CurrentSimNanos**2,
+            "b": lambda CurrentSimNanos: CurrentSimNanos**3,
+        })
+
+        # Run simulation
+
+        times        = log.times()
+        timesSquared = log.a
+        timesCubed   = log.b
+    """
+
+    def __init__(
+        self, 
+        logging_functions: Dict[str, LoggingFunction], 
+        min_log_period: int = 0
+    ) -> None:
+        """Initializer.
+
+        Args:
+            logging_functions (Dict[str, LoggingFunction]): A dictionary where the
+                keys are the names of variables to store, and the values are functions
+                called to retrieve these variables. These functions must accept a single
+                input, an integer with the time in nanoseconds when the function is 
+                called.
+            min_log_period (int, optional): The minimum interval between data recordings
+                Defaults to 0.
```

Comment:
See next comment.

Reply:
Thanks for the review! The `min_log_period` is used here:
```
def UpdateState(self, CurrentSimNanos):
        if CurrentSimNanos >= self._next_update_time:
            # LOG VARIABLE
            self._next_update_time += self.min_log_period
```
As you can see, if `min_log_period==0`, then `_next_update_time` is always 0 and thus `CurrentSimNanos` is always more or equal `_next_update_time`, thus we log every time `UpdateState` is called.

If we were set it to `1`, then `_next_update_time` would slowly increase, but the end effect would be the same. Moreover, I feel like `min_log_period` of 0 carries more meaning (there is no minimum interval) than 1.

Reply:
Sounds great!

Reply:
Also, because you mentioned how we do it in other places in Basilisk. Look up `Recorder` in `src\architecture\messaging\messaging.h`, the new logging class is based on similar functionality. The default value of `Recorder::timeDiff` is 0.